File: prolog-engine/examples/fibonacci.pl
% Fibonacci Sequence — Demonstrates arithmetic and recursion
% ============================================================

fib(0, 0).
fib(1, 1).
fib(N, R) :- N > 1, N1 is N - 1, N2 is N - 2, fib(N1, R1), fib(N2, R2), R is R1 + R2.

% Try these queries:
% ?- fib(6, X).
%   X = 8
%
% ?- fib(10, X).
%   X = 55

File: forth-interpreter/examples/fibonacci.fs
\ Fibonacci sequence using DO/LOOP and recursion
\ Also demonstrates .R for formatted output

: FIB
    DUP 2 < IF EXIT THEN
    DUP 1 - RECURSE SWAP 2 - RECURSE + ;

\ Print a table of Fibonacci numbers
: .FIB-TABLE
    ." n  fib(n)" CR
    ." -------" CR
    20 0 DO
        I 3 .R 2 SPACES
        I FIB 10 .R CR
    LOOP ;

.FIB-TABLE

\ Iterative Fibonacci (more efficient)
: FIB-ITER
    0 1 ROT 0 DO
        OVER + SWAP
    LOOP
    SWAP DROP ;

." Iterative fib(30): " 30 FIB-ITER . CR